src/libraries/Common/src/System/Net/Security/NegotiateStreamPal.Windows.cs

```diff
@@ -328,216 +328,120 @@ internal static Win32Exception CreateExceptionFromError(SecurityStatusPal status
-
-                resultSize += paddingBuffer->cbBuffer;
-                BinaryPrimitives.WriteInt32LittleEndian(output, resultSize);
-
-                return resultSize + 4;
-            }
-        }
-
-        internal static unsafe int Decrypt(
-            SafeDeleteContext securityContext,
-            Span<byte> buffer,
-            bool isConfidential,
-            bool isNtlm,
-            out int newOffset)
-        {
-            if (isNtlm)
-            {
-                return DecryptNtlm(securityContext, buffer, isConfidential, out newOffset);
             }
-
-            //
-            // Kerberos and up
-            //
-            fixed (byte* bufferPtr = buffer)
+            finally
             {
-                Interop.SspiCli.SecBuffer* unmanagedBuffer = stackalloc Interop.SspiCli.SecBuffer[2];
-                Interop.SspiCli.SecBuffer* streamBuffer = &unmanagedBuffer[0];
-                Interop.SspiCli.SecBuffer* dataBuffer = &unmanagedBuffer[1];
-                streamBuffer->BufferType = SecurityBufferType.SECBUFFER_STREAM;
-                streamBuffer->pvBuffer = (IntPtr)bufferPtr;
-                streamBuffer->cbBuffer = buffer.Length;
-                dataBuffer->BufferType = SecurityBufferType.SECBUFFER_DATA;
-                dataBuffer->pvBuffer = IntPtr.Zero;
-                dataBuffer->cbBuffer = 0;
-
-                Interop.SspiCli.SecBufferDesc sdcInOut = new Interop.SspiCli.SecBufferDesc(2)
-                {
-                    pBuffers = unmanagedBuffer
-                };
-
-                uint qop;
-                int errorCode = GlobalSSPI.SSPIAuth.DecryptMessage(securityContext, ref sdcInOut, out qop);
-
-                if (errorCode != 0)
+                if (refAdded)
                 {
-                    Exception e = new Win32Exception(errorCode);
-                    if (NetEventSource.Log.IsEnabled()) NetEventSource.Error(null, e);
-                    throw new Win32Exception(errorCode);
+                    securityContext.DangerousRelease();
                 }
-
-                if (qop == Interop.SspiCli.SECQOP_WRAP_NO_ENCRYPT && isConfidential)
-                {
-                    Debug.Fail($"Expected qop = 0, returned value = {qop}");
-                    throw new InvalidOperationException(SR.net_auth_message_not_encrypted);
-                }
-
-                if (dataBuffer->BufferType != SecurityBufferType.SECBUFFER_DATA)
-                {
-                    throw new InternalException(dataBuffer->BufferType);
-                }
-
-                Debug.Assert((nint)dataBuffer->pvBuffer >= (nint)bufferPtr);
-                Debug.Assert((nint)dataBuffer->pvBuffer + dataBuffer->cbBuffer <= (nint)bufferPtr + buffer.Length);
-                newOffset = (int)((byte*)dataBuffer->pvBuffer - bufferPtr);
-                return dataBuffer->cbBuffer;
             }
         }
 
-        private static unsafe int DecryptNtlm(
+        internal static unsafe void GetMIC(
```

Comment:
Would this produce same output e.g. would it be compatible with other versions of NegotiateStream?

Reply:
Yes, the NTLM code should be compatible.

(There's a part of code where I fixed the on-wire transmission of error codes. It was broken ever since Unix support was added and didn't follow the spec or the .NET Framework version.)

Reply:
it is Kerberos what worries me more. And `SqlClient` uses fragments of this as well. (https://github.com/dotnet/SqlClient/issues/303)

cc: @JRahnama @DavoudEshtehari for visibility

Reply:
There should not be any change of the on-wire data with this PR with two exceptions:
- The fixed error codes that now match the spec and .NET Framework
- Write sends the frame length separately from the data to the underlying transport stream (valid according to `Stream` contract but may be problematic if someone was using `NegotiateStream` to implement `Negotiate` algorithm, and stripping the frame headers)

There's no modification of the native shim beyond what was already done in .NET 7.